src/cryptonote_core/service_node_list.h

```diff
@@ -59,26 +59,40 @@ namespace service_nodes
     void init();
     bool validate_miner_tx(const crypto::hash& prev_id, const cryptonote::transaction& miner_tx, uint64_t height, int hard_fork_version, uint64_t base_reward);
 
-    std::vector<cryptonote::account_public_address> get_expired_nodes(uint64_t block_height) const;
-    cryptonote::account_public_address select_winner(const crypto::hash& prev_id);
+    std::vector<crypto::public_key> get_expired_nodes(uint64_t block_height) const;
+
+    std::vector<std::pair<cryptonote::account_public_address, uint32_t>> get_winner_addresses_and_shares(const crypto::hash& prev_id) const;
+    crypto::public_key select_winner(const crypto::hash& prev_id) const;
 
     bool is_service_node(const crypto::public_key& pubkey) const;
     const std::shared_ptr<quorum_state> get_quorum_state(uint64_t height) const;
 
   private:
 
-    bool is_registration_tx(const cryptonote::transaction& tx, uint64_t block_height, cryptonote::account_public_address& address, crypto::public_key& key) const;
-    bool is_deregistration_tx(const cryptonote::transaction& tx, cryptonote::account_public_address& address) const;
+    struct service_node_info
+    {
+      // Service nodes are organized by time since last reward or registration
+      // This value is given by block height, and differentiated by transaction index for
+      // registrations that occured in the same block. index = 0 for block reward, 1 for first transaction, etc.
+      // hence a std::pair<uint64_t, size_t> is used here for this value.
+      std::pair<uint64_t, size_t> height;
```

Comment:
stl pairs already have a operator< defined. This is clearer for me, sorry.

Reply:
I find this easier to read because I'm using them as an index, and STL pairs have operator< defined as indexed on the first, then the second element.

I'll go ahead and change it here though and construct the pairs on the fly when I do the comparison.

Reply:
Much nicer now!

Reply:
Fair enough, changed